### PR TITLE
Expand Javadocs on writeIndexGen

### PR DESCRIPTION
**Pull request by DaveCTurner**
Adds a little more context describing why we use the three-step process,
and a note about the meaning of a failure at this step.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -2142,12 +2142,27 @@ public boolean isReadOnly() {
...
+     * We use this process to protects against problems such as a master failover part-way through. If a new master is elected while we're
+     * writing the root blob with generation {@code P} then we will fail to update the safe repository generation in the final step, and
+     * meanwhile the new master will choose a greater generation for all subsequent root blobs avoiding any danger of clobbering each
+     * other's writes.
+     * <p>
+     * Note that a failure here does not imply that the process was unsuccessful or the repository is unchanged. Once we have written the
+     * new root blob the repository is updated from the point of view of any other clusters reading from it, and if we performed a full
+     * cluster restart at that point then we would also pick up the new root blob. Writing the root blob may succeed without us receiving
+     * a successful response from the repository, leading us to report that the write operation failed. Updating the safe generation may
+     * likewise succeed on a majority of master-eligible nodes which does not include this one, again leading to an apparent failure.
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2165]: IIUC, this means some later failures from this method do not mean the whole snapshot has actually failed. But there are still eariler failures, e.g. failure in step 1, that do lead to actual failure? Not sure if it is possible, but if master changed before step 1 can happen, files written in this snapshot process will become dangling?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -2142,12 +2142,27 @@ public boolean isReadOnly() {
...
-     * First, the {@link RepositoryMetadata} entry for this repository is set into a pending state by incrementing its
-     * pending generation {@code P} while its safe generation {@code N} remains unchanged.
-     * Second, the updated {@link RepositoryData} is written to generation {@code P + 1}.
-     * Lastly, the {@link RepositoryMetadata} entry for this repository is updated to the new generation {@code P + 1} and thus
-     * pending and safe generation are set to the same value marking the end of the update of the repository data.
+     * Writing a new index generation (root) blob is a three-step process:
+     * <ol>
+     * <li>
+     * The {@link RepositoryMetadata} entry for this repository is set into a pending state by increasing its pending generation {@code P}
+     * while its safe generation {@code N < P} remains unchanged.
```
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2149]: Can we somehow make it more explicit that N and P begin as equal numbers and throughout their lifecycles, N will always be less than or equal to P?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2165]: That's true, but there's not really a reliable way to distinguish this case from the caller's point of view. A failure in step 1 will prevent us attempting to write a new root blob but may still have changed the pending generation (on a majority of masters, which is what matters).

Leaks here will be cleaned up by subsequent operations on the repository anyway.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I added some more details on this in 8c798f9ad53005996b5e1cdf19345745d07c0b1d. It's not true that N and P always start out equal, although that's typically the case.

**Review by ywangd — APPROVED** [review]
LGTM 

Thanks for the explanation!

**Pull request merged**